stan/math/rev/core/vari.hpp

```diff
@@ -179,11 +165,375 @@ class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
   static inline void operator delete(
       void* /* ignore arg */) noexcept { /* no op */
   }
+
+ private:
+  template <typename>
+  friend class var_value;
 };
 
 // For backwards compatability the default is double
 using vari = vari_value<double>;
 
+/**
+ * The variable implementation for Eigen dense matrix types.
+ *
+ * This class is complete (not abstract) and may be used for
+ * constants.
+ *
+ * A variable implementation is constructed with a constant
+ * value. It also stores the adjoint for storing the partial
+ * derivative with respect to the root of the derivative tree.
+ *
+ */
+template <typename T>
+class vari_value<T, require_eigen_dense_base_t<T>> : public vari_base {
+ public:
+  static_assert(
+      is_plain_type<T>::value,
+      "The template for this var is an"
+      " expression but a var_value's inner type must be assignable such as"
+      " a double, Eigen::Matrix, or Eigen::Array");
+
+  /**
+   * `PlainObject` represents a user constructible type such as Matrix or Array
+   */
+  using PlainObject = std::decay_t<plain_type_t<T>>;
+  using Scalar = PlainObject;  // The underlying type for this class
+  using value_type = Scalar;   // The underlying type for this class
+  using eigen_scalar = value_type_t<PlainObject>;  // A floating point type
```

Comment:
I'm fine with making it an actual scalar type. @t4c1 made `scalar_type_t` return a `var<double>` for `var<matrix>` so I'm fine with making this a `var<double>`

Reply:
Or should it be a `double`?

Reply:
I think what @t4c1 did makes sense if we want this to be a scalar_type (which maybe we do so some metaprograms work?).

I could also see a matrix-something making sense if we wanted to rename this to something else.

This is one of those issues I'm not gonna be able to understand fully until I write code (to know which thing is useful).

Reply:
Yeah it's a really confusing area because this class can be both a `Scalar` type on it's own or a container. I think I like `Scalar = double`. If you want to query whether it's a container / it's container type you can look at the `value_type` but if you just want to know "Is this a `double, float, long double` floating point type" you can query the `Scalar`

Reply:
Well I suspect the uses of a scalar_type here would be to determine if things are autodiff or not or build a return type. In that case I want the var-ish-ness to come through.

So I'm also a little shaky nailing it down until we have the code that uses this.

Reply:
Should I just get rid of it till we want it?